Lib/tarfile.py

```diff
@@ -546,26 +547,27 @@ def _read(self, size):
                 buf = self.cmp.decompress(buf)
             except self.exception:
                 raise ReadError("invalid compressed data")
-            self.dbuf += buf
+            t.append(buf)
             c += len(buf)
-        buf = self.dbuf[:size]
-        self.dbuf = self.dbuf[size:]
-        return buf
+        t = b"".join(t)
+        self.dbuf = t[size:]
+        return t[:size]
 
     def __read(self, size):
         """Return size bytes from stream. If internal buffer is empty,
            read another block from the stream.
         """
         c = len(self.buf)
+        t = [self.buf]
```

Comment:
I don't think this optimization is needed.
In all caller of `__read()`, `size` is small or `==bufsize`.
`while c < size:` doesn't loop twice actually.

Reply:
For large `size`, chunked read is not needed anyway.

```py
rem = size - len(self.buf)
if rem > 0:
    self.buf += self.fileobj.read(max(self.bufsize, rem))
t = self.buf[:size]
self.buf = self.buf[size:]
return t
```